engine/table/src/main/java/io/deephaven/engine/table/impl/select/analyzers/SelectAndViewAnalyzer.java

```diff
@@ -55,32 +58,50 @@ public static SelectAndViewAnalyzer create(final Mode mode, final Map<String, Co
             rowRedirection = null;
         }
 
-        final boolean flatResult = rowSet.isFlat();
+        final TrackingRowSet originalRowSet = rowSet;
+        boolean flatResult = rowSet.isFlat();
         // if we preserve a column, we set this to false
         boolean flattenedResult = !flatResult
                 && allowInternalFlatten
                 && (columnSources.isEmpty() || !publishTheseSources)
                 && mode == Mode.SELECT_STATIC;
         int numberOfInternallyFlattenedColumns = 0;
 
+        final HashSet<String> resultColumns = flattenedResult ? new HashSet<>() : null;
         for (final SelectColumn sc : selectColumns) {
-            final Map<String, ColumnSource<?>> columnsOfInterest = analyzer.getAllColumnSources();
             analyzer.updateColumnDefinitionsFromTopLayer(columnDefinitions);
             sc.initDef(columnDefinitions);
-            sc.initInputs(rowSet, columnsOfInterest);
+            sc.initInputs(rowSet, analyzer.getAllColumnSources());
+
+            // When flattening the result, intermediate columns generate results in position space. When we discover
+            // that a select column depends on an intermediate result, then we must flatten all parent columns so
+            // that all dependent columns are in the same result-key space.
+            if (!flatResult && flattenedResult && Stream.concat(sc.getColumns().stream(), sc.getColumnArrays().stream())
+                    .anyMatch(resultColumns::contains)) {
+                analyzer = analyzer.createStaticFlattenLayer(rowSet);
+                rowSet = RowSetFactory.flat(rowSet.size()).toTracking();
+                flatResult = true;
+
+                // we must re-initialize the column inputs as they may have changed post-flatten
+                sc.initInputs(rowSet, analyzer.getAllColumnSources());
+            } else if (!flatResult && flattenedResult) {
+                resultColumns.add(sc.getName());
+            }
+
             final Stream<String> allDependencies =
                     Stream.concat(sc.getColumns().stream(), sc.getColumnArrays().stream());
             final String[] distinctDeps = allDependencies.distinct().toArray(String[]::new);
             final ModifiedColumnSet mcsBuilder = new ModifiedColumnSet(parentMcs);
 
             if (shouldPreserve(sc)) {
-                analyzer =
-                        analyzer.createLayerForPreserve(sc.getName(), sc, sc.getDataView(), distinctDeps, mcsBuilder);
                 if (numberOfInternallyFlattenedColumns > 0) {
```

Comment:
How are we updating numberOfInternallyFlattenedColumns on line 81 so that we will bail out of the flattening to allow us to preserve other in-memory column sources?

Reply:
I get it now.  We had to have at least one result column that was previously flattened to get into the case on line 81.  We'll flatten out the rest of our internal columns at that point.  If we end up hitting shouldPreserve later, then having one is enough to restart the analysis.